### ISSUE TITLE
Running C++ tests

### ISSUE THREAD

**Issue by kotp** [exercism/xcpp#115]
@DolevEfr commented on [Sun Mar 26 2017](https://github.com/exercism/exercism.io/issues/3436)

Hello good people, 
I started out with the C++ exercises, and I must say that your documentation is really good for setting everything up (at least on linux).
The only thing missing there (and in the readme) is how to run the tests...
I'm new to C++ and developing using the terminal, and I can't figure out how to run the tests.

The first time I run 'make'- it gave me compile-time errors in the tests file, but once I finished declaring the essential namespaces and functions- it just compiles my binary without giving me any feedback regarding the tests (I know for sure I don't pass the first test).

---

@kytrinyx commented on [Sun Mar 26 2017](https://github.com/exercism/exercism.io/issues/3436#issuecomment-289295226)

@DolevEfr Thanks! We should absolutely see about getting that fixed. Would you mind opening an issue in the https://github.com/exercism/xcpp repository? That should get this in front of the people who are best suited to take care of this.

There's some information here: http://exercism.io/languages/cpp/tests

I suggest using the 'HINTS.md' file in the first exercise to help document how to run the tests (or where to find the information about running the tests). The documentation about how to update the exercise READMEs is here: https://github.com/exercism/docs/blob/master/contributing-to-language-tracks/fixing-exercise-readmes.md

**Comment by patricksjackson** [comment]
Hmm, do you get any output when you run make? Does it create an output executable (should be the same as the exercise name)?

The tests work by creating an executable that will just runs the test cases and reports the output. When you run `make` to build the tests, it will autorun that executable as part of the build process. So you should get a couple lines of output while Make compiles the files, then the output from the tests. Something like this:

```
Scanning dependencies of target hello-world
[ 50%] Building CXX object CMakeFiles/hello-world.dir/hello_world_test.cpp.o
[100%] Linking CXX executable hello-world
Running 1 test case...

*** No errors detected
[100%] Built target hello-world
```

One thing it could be, it won't show anything if you haven't modified any of the files since the last time you ran make (it will just report `[100%] Built target hello-world`). If that's the case you can run `make clean && make` which will force it to recompile.

**Comment by lverns** [comment]
I also am confused about how to run the tests. 

I'm on NixOS 17.03.

### Reproducing

- Make sure the environment is clean: `rm -rf ~/exercism/cpp/hello-world/`
- Check out the exercise `exercism fetch cpp hello-world`
- `cd ~/exercism/cpp/hello-world/`
- Change `set(Boost_USE_STATIC_LIBS ON)` to `set(Boost_USE_STATIC_LIBS OFF)` in `CMakeLists.txt` (see related issues #116 and #123)
- `mkdir build`
- `cd build`
- `cmake -G "Unix Makefiles" ..` (completes with no errors)
- `cd ..`
- Create `hello_world.cpp` with the following contents
```C++
#include <string>

namespace hello_world {
  std::string hello() {
    return "Hello World";
  }
}
```
- `cd build && make` creates the following output
```
[ 33%] Building CXX object CMakeFiles/hello-world.dir/hello_world_test.cpp.o
[ 66%] Building CXX object CMakeFiles/hello-world.dir/hello_world.cpp.o
[100%] Linking CXX executable hello-world
/nix/store/63gvnrj4z154kpyjpskl6s0hwmyx9x0w-glibc-2.25/lib/crt1.o: In function `_start':
/tmp/nix-build-glibc-2.25.drv-0/glibc-2.25/csu/../sysdeps/x86_64/start.S:110: undefined reference to `main'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/hello-world.dir/build.make:124: hello-world] Error 1
make[1]: *** [CMakeFiles/Makefile2:68: CMakeFiles/hello-world.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```
- So I tried changing `hello_world.cpp` to be the following:
```C++
#include <string>

int main() {
  return 0;
}

namespace hello_world {
  std::string hello() {
    return "Hello World";
  }
}

```
- Running `make` again now gives the following output
```
[ 33%] Building CXX object CMakeFiles/hello-world.dir/hello_world.cpp.o
[ 66%] Linking CXX executable hello-world
[100%] Built target hello-world
```
- I can make the string returned by the `hello()` method be any string and I still get the same output.

**Comment by lverns** [comment]
Doing the same steps on Fedora 24 produces roughly the same results up until I try adding `main()` to `hello_world.cpp`.

**Comment by patricksjackson** [comment]
@lverns - The tests work by letting Boost provide a main() function that will run all the tests we've defined. (That config is set by defining `BOOST_TEST_MAIN` in the *_test.cpp file.) Boost needs to do things a little bit differently if you are compiling against a shared lib rather than a static lib, so you need to let it know you are doing that. The script in #116 makes that change for you.

TL;DR - Just add the line `#define BOOST_TEST_DYN_LINK` to the top of the *_test.cpp file.

**Comment by lverns** [comment]
@patricksjackson - Thanks! That solves the problem.

```
[ 33%] Building CXX object CMakeFiles/hello-world.dir/hello_world.cpp.o
[ 66%] Linking CXX executable hello-world
Running 1 test case...

*** No errors detected
[100%] Built target hello-world

```
